controllers/storagecluster_controller.go

```diff
@@ -45,17 +58,111 @@ type StorageClusterReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *StorageClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	instance := &ocsv1.StorageCluster{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("storagecluster instance not found")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("storagecluster instance found")
+
+	// get list of StorageSystems
+	storageSystemList := &odfv1alpha1.StorageSystemList{}
+	err = r.Client.List(context.TODO(), storageSystemList, &client.ListOptions{Namespace: instance.Namespace})
+	if err != nil {
+		logger.Error(err, "failed to list the storageSystem")
+		return ctrl.Result{}, err
+	}
+
+	storageSystem := filterStorageSystem(storageSystemList, instance.Name)
+	if storageSystem == nil {
+		logger.Info("StorageSystem not found for the storageCluster, will create one")
+
+		storageSystem = &odfv1alpha1.StorageSystem{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      instance.Name,
+				Namespace: instance.Namespace,
+			},
+			Spec: odfv1alpha1.StorageSystemSpec{
+				Name:      instance.Name,
+				NameSpace: instance.Namespace,
+				Kind:      odfv1alpha1.StorageCluster,
+			},
+		}
+
+		// create StorageSystem for storageCluster
+		err = r.Client.Create(context.TODO(), storageSystem)
+		if err != nil {
+			logger.Error(err, "failed to create StorageSystem", "StorageSystem", storageSystem.Name)
+			return ctrl.Result{}, err
+		}
+		logger.Info("StorageSystem created for the storageCluster")
+	} else {
+		logger.Info("StorageSystem is already present for the storageCluster", "StorageSystem", storageSystem.Name)
+	}
 
-	// your logic here
+	// update Annotation for the storageCluster
+	metav1.SetMetaDataAnnotation(&instance.ObjectMeta, string(HasStorageSystem), "true")
+	err = r.Client.Update(context.TODO(), instance)
+	if err != nil {
+		logger.Error(err, "failed to update the storageCluster Annotation")
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func filterStorageSystem(storageSystemList *odfv1alpha1.StorageSystemList, storageClusterName string) *odfv1alpha1.StorageSystem {
+
+	for _, ss := range storageSystemList.Items {
+		if ss.Spec.Name == storageClusterName && ss.Spec.Kind == odfv1alpha1.StorageCluster {
```

Comment:
Should `odfv1alpha1.StorageCluster` be `ocsv1.StorageCluster`?

Reply:
no `odfv1alpha1.StorageCluster` is just a string and not the whole type like `ocsv1.StorageCluster`

https://github.com/red-hat-data-services/odf-operator/blob/main/api/v1alpha1/storagesystem_types.go#L37

Reply:
....why are we defining the StorageCluster Kind as a string in odf-operator? This seems backwards. Shouldn't we be referencing or generating it from somewhere in `ocsv1`?

Reply:
We can do that, But I want to keep control of these variables in my own hands So that we can keep them the way we want instead of asking the vendors to do that.

for example it was `StorageCluster StorageKind = "StorageCluster"` and we changed it to the `StorageCluster StorageKind = "storagecluster.ocs.openshift.io/v1"`

Moreover, this is a variable for `kind` in our `spec`. For which we do need the same type to define all the vendors. https://github.com/red-hat-data-services/odf-operator/blame/main/api/v1alpha1/storagesystem_types.go#L60